### ISSUE TITLE
[Request] Record video in Colab

### ISSUE THREAD

**Issue by Charelvanhoof** [hill-a/stable-baselines#78]
First of all, excellent work, just what I have been looking for!
Love it that I can experiment and have well documented code.

I work mainly in Google Colab, mainly because I got tired of pip install <whatever> and it removes the hassle of infrastructure.

See you published very nice Colabs, many thanks.

I figured out a way with wrappers to have Gym create video's that is also working in Colab (with virtual monitor) , nice to have a graph explaining the score is increasing but also want to see games with my own eyes. Unfortunately that does not work anymore with Stable baselines (at least I did not yet spot it).

Any ideas how to solve this for a Colab environments with Stable-baselines?

Below example how it works for e in a standard gym environment:
`
# @title Set-up the virtual display environment
!apt-get update
!apt-get install python-opengl -y
!apt install xvfb -y
!pip install pyvirtualdisplay
!pip install piglet
!apt-get install ffmpeg

# @title Start the virtual monitor
from pyvirtualdisplay import Display
display = Display(visible=0, size=(1400, 900))
display.start()

# @title play a random game and create video
env = gym.make("PongNoFrameskip-v4")
monitor_dir ='/content/test5'

#Setup a wrapper to be able to record a video of the game
record_video = True
should_record = lambda i: record_video
env = wrappers.Monitor(env, monitor_dir, video_callable=should_record, force=True)

#Play a game
state = env.reset()
done = False
while not done:
  action = env.action_space.sample() #random action, replace by the prediction of the model
  state, reward, done, _ = env.step(action)

record_video = False
env.close() 

# download videos
from google.colab import files
import glob
os.chdir(monitor_dir) # change directory to get the files
!pwd #show file path
!ls # show directory content

for file in glob.glob("*.mp4*"):
  print(file)
  files.download(file)

os.chdir('..') #backout the video directory
`

Wrapping the whole stable-baseline env does not work.

Tried also something like:

`
monitor_dir ='/content/test7'
env_show = gym.make('LunarLander-v2')
record_video = True
should_record = lambda i: record_video
env_show = wrappers.Monitor(env_show, monitor_dir, video_callable=should_record, force=True)

obs = env_show.reset()
dones = False
while not dones:
    action, _states = model.predict(obs)
    obs, rewards, dones, info = env_show.step(action)
    env_show.render()

record_video = False
env_show.close() 
`
Which works for Lunar Lander but not for an Atari game: Error: Unexpected observation shape (210, 160, 3) for Box environment, please use (84, 84, 4) or (n_env, 84, 84, 4) for the observation shape.

Any pragmatic elegant solution or small enhancement for Colab users?

Thank you

**Comment by araffin** [comment]
Hi, 
Thanks you for your request.
It seems that an important part is missing: the Monitor class that you are using.
I would also use VideoRecorder from gym (see https://github.com/openai/gym/blob/master/gym/wrappers/monitoring/video_recorder.py ). To work with VecEnc, I would advise you to take a look at what was done in the baselines repo here: https://github.com/openai/baselines/blob/master/baselines/common/vec_env/vec_video_recorder.py

Also, please use markdown code blocks (see https://guides.github.com/features/mastering-markdown/):
```
```python
def test():
    pass
\```
```

**Comment by Charelvanhoof** [comment]
Many thanks for the quick reply, will try it out. 
Note: I used the markdown button "insert code" of the website but indeed did not give the wanted markdown but only single '', apologies I did not spot it.

Suggestion to add a ready-to-use colab example to you repository and on your website to record a video and download the result, I'll bet you you will reach more people.

**Comment by araffin** [comment]
>Suggestion to add a ready-to-use colab example to you repository and on your website to record a >video and download the result, I'll bet you you will reach more people. The updated "enjoy script" to >record a video in colab

Well, if you provide such a notebook, we will certainly link it in the documentation ;)

**Comment by Charelvanhoof** [comment]
:-)
I am trying, but I am afraid my skills in this domain are not as good as one would hope.

I first tried your first link:

'''
# Enjoy trained agent
from gym.wrappers.monitoring.video_recorder import VideoRecorder

monitor_dir ='/content/test12'
env = VideoRecorder(env, base_path=monitor_dir, enabled=True)

obs = env.reset()
dones=False
while not dones:
    action, _states = model.predict(obs)
    obs, rewards, dones, info = env.step(action)
    env.render()

VideoRecorder.close() 
'''
which got me the error: AttributeError: 'VecFrameStack' object has no attribute 'metadata'

So I have to search further in the second link.
Any suggestions most welcome :-)

**Comment by araffin** [comment]
The video recorder provided by gym will only work with simple gym env, not VecEnv nor VecEnvWrapper (that are used by stable-baselines to facilitate multiprocessing and preprocessing)

**Comment by araffin** [comment]
@Charelvanhoof I have got a good news for you. I have been tweaking VecVideoRecorder so it should work with any type of VecEnv.
It is available on the 'record-video-branch', here: https://github.com/hill-a/stable-baselines/blob/record-videos/stable_baselines/common/vec_env/vec_video_recorder.py

Consider this feature to be experimental

a minimal example:
```python
import gym
from stable_baselines.common.vec_env import VecVideoRecorder, DummyVecEnv

env_id = 'CartPole-v1'
video_folder = 'logs/videos/'
video_length = 100

env = DummyVecEnv([lambda: gym.make(env_id)])

obs = env.reset()

env = VecVideoRecorder(env, video_folder,
                       record_video_trigger=lambda x: x == 0, video_length=video_length,
                       name_prefix="random-agent-{}".format(env_id))

env.reset()
for _ in range(video_length + 1):
    action = [env.action_space.sample()]
    obs, _, _, _ = env.step(action)
env.close()

```

Important note: ffmpeg or avconc must be installed on the machine

**Comment by Charelvanhoof** [comment]
@araffin brilliant, than you! I have tested this morning with various environments  (eg SpaceInvaders, Pong, etc) and it works. Also with make_atari_env and VecFrameStack.

I vote this feature to be upgraded from experimental to baseline.

An example of training an agent (also to check I did it right), e.g. is the 
```
action, _states = model.predict(obs)
```
the correction interpretation because was not sure about the [ ] brackets in the example. 

Below a lunar lander with training:
```
# @title trained agent
env_id = 'LunarLander-v2'
monitor_dir = 'logs/videos/'
video_length = 300

env = DummyVecEnv([lambda: gym.make(env_id)])

model = A2C(MlpPolicy, env, ent_coef=0.1, verbose=1)
model.learn(total_timesteps=35000)

print("Training complete")

obs = env.reset()

env = VecVideoRecorder(env, monitor_dir,
                       record_video_trigger=lambda x: x == 0, video_length=video_length,
                       name_prefix="trained-agent-{}".format(env_id))

env.reset()
for _ in range(video_length + 1):
    action, _states = model.predict(obs)
    obs, _, _, _ = env.step(action)
env.close()
```

Or pong with training:

```
env_id = 'PongNoFrameskip-v4'
monitor_dir = 'logs/Pong/'
video_length = 1000

# There already exists an environment generator
# that will make and wrap atari environments correctly.
# Here we are also multiprocessing training (num_env=4 => 4 processes)
env = make_atari_env(env_id, num_env=4, seed=0)
# Frame-stacking with 4 frames
env = VecFrameStack(env, n_stack=4)

model = ACER(CnnPolicy, env, verbose=1)
model.learn(total_timesteps=50000)

print("Training complete")

obs = env.reset()

env = VecVideoRecorder(env, monitor_dir,
                       record_video_trigger=lambda x: x == 0, video_length=video_length,
                       name_prefix="trained-agent-{}".format(env_id))

env.reset()
for _ in range(video_length + 1):
    #action = [env.action_space.sample()]
    action, _states = model.predict(obs)
    obs, _, _, _ = env.step(action)
env.close()
```
Here you notice the video has 4 screens pong in parallel, kind of nice as well.

**Comment by araffin** [comment]
>I have tested this morning with various environments (eg SpaceInvaders, Pong, etc) and it works. Also >with make_atari_env and VecFrameStack.
Yes, in fact I used [this script](https://github.com/araffin/rl-baselines-zoo/blob/video-recorder/utils/record_video.py) from https://github.com/araffin/rl-baselines-zoo to test the feature ;)

>I vote this feature to be upgraded from experimental to baseline.

I created the pull request, waiting for @hill-a review.

**Comment by araffin** [comment]
Yes, with vec env, actions must be array even with only one env, hence the []
